chasm/lib/scheduler/scheduler.go

```diff
@@ -280,13 +296,174 @@ func (s *Scheduler) recordActionResult(result *schedulerActionResult) {
 	}
 
+	// Update RunningWorkflows.
 	for _, start := range result.starts {
-		if start.StartWorkflowResult != nil {
+		completed := start.StartWorkflowStatus >= enumspb.WORKFLOW_EXECUTION_STATUS_COMPLETED
+
+		if !completed {
+			// Add running workflows to the list (duplicates are acceptable for ALLOW_ALL policy)
 			s.Info.RunningWorkflows = append(s.Info.RunningWorkflows, start.StartWorkflowResult)
+		} else {
+			// Remove completed workflows by WorkflowId+RunId
+			wid := start.StartWorkflowResult.WorkflowId
+			rid := start.StartWorkflowResult.RunId
+			runningIdx := slices.IndexFunc(s.Info.RunningWorkflows, func(wfe *commonpb.WorkflowExecution) bool {
+				return wfe.WorkflowId == wid && wfe.RunId == rid
+			})
+			if runningIdx >= 0 {
+				s.Info.RunningWorkflows = slices.Delete(s.Info.RunningWorkflows, runningIdx, runningIdx+1)
+			}
 		}
 	}
 }
+
+var _ chasm.NexusCompletionHandler = &Scheduler{}
+
+func executionStatusFromFailure(failure *failurepb.Failure) enumspb.WorkflowExecutionStatus {
+	switch failure.FailureInfo.(type) {
+	case *failurepb.Failure_CanceledFailureInfo:
+		return enumspb.WORKFLOW_EXECUTION_STATUS_CANCELED
+	case *failurepb.Failure_TimeoutFailureInfo:
+		return enumspb.WORKFLOW_EXECUTION_STATUS_TIMED_OUT
+	default:
+		return enumspb.WORKFLOW_EXECUTION_STATUS_FAILED
+	}
+}
+
+// HandleNexusCompletion allows Scheduler to record workflow completions from
+// worfklows started by the same scheduler tree's Invoker.
+func (s *Scheduler) HandleNexusCompletion(
+	ctx chasm.MutableContext,
+	info *persistencespb.ChasmNexusCompletion,
+) error {
+	invoker, err := s.Invoker.Get(ctx)
+	if err != nil {
+		return err
+	}
+
+	workflowID := invoker.WorkflowID(info.RequestId)
+	if workflowID == "" {
+		// If the request ID was removed, the request must have already been processed;
+		// fast-succeed.
+		return nil
+	}
+
+	// Handle last completed/failed status and payloads.
+	//
+	// TODO - also record payload sizes once we have metrics wired into CHASM context.
+	var wfStatus enumspb.WorkflowExecutionStatus
+	switch outcome := info.Outcome.(type) {
+	case *persistencespb.ChasmNexusCompletion_Failure:
+		wfStatus = executionStatusFromFailure(outcome.Failure)
+		s.LastCompletionResult = chasm.NewDataField(ctx, &schedulerpb.LastCompletionResult{
+			Outcome: &schedulerpb.LastCompletionResult_Failure{
+				Failure: outcome.Failure,
+			},
+		})
+	case *persistencespb.ChasmNexusCompletion_Success:
+		wfStatus = enumspb.WORKFLOW_EXECUTION_STATUS_COMPLETED
+		s.LastCompletionResult = chasm.NewDataField(ctx, &schedulerpb.LastCompletionResult{
+			Outcome: &schedulerpb.LastCompletionResult_Success{
+				Success: outcome.Success,
+			},
+		})
+	default:
+		wfStatus = enumspb.WORKFLOW_EXECUTION_STATUS_FAILED
+	}
+
+	// Handle pause-on-failure.
+	if wfStatus != enumspb.WORKFLOW_EXECUTION_STATUS_COMPLETED &&
+		s.Schedule.Policies.PauseOnFailure && !s.Schedule.State.Paused {
```

Comment:
Same here, do you always want to pause? Even if the workflow is a backfill or not the most recent one?